README.md
# picasso

## Introduction

`picasso` is a PICA200 shader assembler, written in C++. The PICA200 is the GPU used by the Nintendo 3DS.

`picasso` comes with a manual `Manual.md` that explains the shader language. `example.vsh` is simple example that demonstrates it.

## Building

A working C++ compiler for the host is required (Windows users: use TDM-GCC), plus autotools. Use the following commands to build the program:

    ./autogen.sh
    ./configure
    make

## Shout-outs

- **smea** for reverse-engineering the PICA200, writing documentation, working hard & making `aemstro_as.py` (the original homebrew PICA200 shader assembler)
- **neobrain** for making `nihstro-assemble`, whose syntax inspired that of `picasso` and whose usage of boost inspired me to make my own assembler without hefty dependencies.


Changelog.md
# picasso Changelog

# v2.7.1

- Further improvements to overall system stability and other minor adjustments have been made to enhance the user experience.

# v2.7

- Added `dst`, `litp` and `break` instructions (thanks to @Tilka).
- Added check to enforce index regs being used only with floating point vector uniforms.
- Renamed index registers to match D3D naming conventions (`a0.x`, `a0.y`, `aL`) (old names still accepted).
- Miscellaneous bugfixes and improvements (thanks to @lioncash).

# v2.6.2

- Fixed several compilation errors in some compilers.

# v2.6.1

- Reduced `mad` opdesc allocation errors by automatically swapping out of bounds opdesc entries with other ones in the addressable range (5 bits).

# v2.6

- Added `.in` directive for explicit specifying (and allocating) input registers and exporting them in the DVLE uniform table.
- Added support for dollar signs (`$`) in identifier names, which are translated to period characters (`.`) in DVLE uniform names.
- Output registers `o7` through `o15` are now allowed in vertex shaders (as dummy outputs).
- DVLE uniform table is now sorted by register position.

# v2.5

- The `.gsh` directive has been enhanced to provide full support for all geometry shader operation modes (point, variable-sized primitive and fixed-size primitive). This also effectively separates vertex shader uniform space from geometry shader uniform space.
- The `.out` directive has been enhanced to allow wiring semantics to any arbitrary output register. Additionally the `dummy` semantic was added while the `7` semantic was removed.
- Added auto-insertion of NOP instruction in corner cases involving flow of control instructions, together with the `--no-nop` directive which instead of adding NOPs warns the user about the corner cases.
- Added support for `rgba` and `stpq` in addition to `xyzw`.
- Added an error message for invalid input register use (e.g. `add r0, v1, v2`).
- The operand descriptor allocation algorithm has been enhanced to take into account unused operands.
- The `6` and `7` conditional operators have been removed since they actually do not exist.
- Really corrected MAD instruction encoding.
- Several miscellaneous issues were fixed.

# v2.4

- Corrected MAD instruction encoding.
- Added command line flag for retrieving the picasso version.

# v2.3

- Added `.constfa` for creating floating-point vector constant arrays.
- Fixed `.nodvle` bug.

# v2.2

- Added proper support for the MOVA instruction.
- Added support for inverting the condition in JMPU.
- Fixed `lcnt` bug.

# v2.1

- Fixed input file open error message.
- Fixed `.constf` misallocation bug.

# v2.0

- (**Breaking change**) Command line format changed.
- Added support for assembling multiple shaders (DVLEs) into a single SHBIN.
- Added new directives: `.entry`, `.nodvle`, `.gsh`, `.setf`, `.seti`, `.setb`.
- Added auto-detection of inverted forms of opcodes. (Explicitly using `dphi`, `sgei`, `slti` and `madi` is now deprecated)
- Several miscellaneous bug fixes.

# v1.0

- Initial release.


source/types.h
#pragma once
#include <stdint.h>

typedef uint64_t dword_t;
typedef uint32_t word_t;
typedef uint16_t hword_t;
typedef uint8_t byte_t;
typedef int64_t dlong_t;
typedef int32_t long_t;
typedef int16_t short_t;
typedef int8_t char_t;
typedef uint64_t u64;
typedef uint32_t u32;
typedef uint16_t u16;
typedef uint8_t u8;

#define BIT(n) (1U << (n))

#ifndef __BYTE_ORDER__
#include <sys/param.h>
#define __BYTE_ORDER__ BYTE_ORDER
#define __ORDER_LITTLE_ENDIAN__ LITTLE_ENDIAN
#define __ORDER_BIG_ENDIAN__ BIG_ENDIAN
#endif

#ifndef __llvm__
#if !defined(__GNUC__) || (__GNUC__ < 4) || (__GNUC__ == 4 && __GNUC_MINOR__ < 8)

static inline uint16_t __builtin_bswap16(uint16_t x)
{
	return ((x << 8) & 0xff00) | ((x >> 8) & 0x00ff);
}

#if defined(__GNUC__) && (__GNUC__ == 4 && __GNUC_MINOR__ < 7)
static inline uint32_t __builtin_bswap32(uint32_t x)
{
	return ((x << 24) & 0xff000000) |
	       ((x <<  8) & 0x00ff0000) |
	       ((x >>  8) & 0x0000ff00) |
	       ((x >> 24) & 0x000000ff);
}

static inline uint64_t __builtin_bswap64(uint64_t x)
{
	return (uint64_t)__builtin_bswap32(x>>32) |
	      ((uint64_t)__builtin_bswap32(x&0xFFFFFFFF) << 32);
}
#endif
#endif
#endif

#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
#define be_dword(a)  __builtin_bswap64(a)
#define be_word(a)  __builtin_bswap32(a)
#define be_hword(a) __builtin_bswap16(a)
#define le_dword(a)  (a)
#define le_word(a)  (a)
#define le_hword(a) (a)
#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
#define be_dword(a)  (a)
#define be_word(a)  (a)
#define be_hword(a) (a)
#define le_dword(a)  __builtin_bswap64(a)
#define le_word(a)  __builtin_bswap32(a)
#define le_hword(a) __builtin_bswap16(a)
#else
#error "What's the endianness of the platform you're targeting?"
#endif


source/picasso.h
#pragma once
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdarg.h>
#include <getopt.h>
#ifdef WIN32
#include <fcntl.h>
#endif
#include "types.h"

#include <vector>
#include <list>
#include <map>
#include <string>
#include <algorithm>

#include "FileClass.h"

#include "maestro_opcodes.h"

#if !defined(WIN32) && !defined(stricmp)
#define stricmp strcasecmp
#endif

enum
{
	COMP_X = 0,
	COMP_Y,
	COMP_Z,
	COMP_W,
};

#define SWIZZLE_COMP(n,v) ((v) << (6-(n)*2))
#define OPSRC_MAKE(neg, sw) ((neg) | ((sw) << 1))
#define OPDESC_MAKE(out, src1, src2, src3) ((out) | ((src1) << 4) | ((src2) << (4+9)) | ((src3) << (4+9*2)))
#define FMT_OPCODE(n) ((n)<<26)
#define OUTPUT_MAKE(i, reg, mask) ((i) | ((reg)<<16) | ((u64)(mask)<<32))

#define DEFAULT_SWIZZLE (SWIZZLE_COMP(0,COMP_X) | SWIZZLE_COMP(1,COMP_Y) | SWIZZLE_COMP(2,COMP_Z) | SWIZZLE_COMP(3,COMP_W))
#define DEFAULT_OPSRC OPSRC_MAKE(0, DEFAULT_SWIZZLE)

#define OPDESC_MASK_D123 OPDESC_MAKE(0xF, 0x1FF, 0x1FF, 0x1FF)
#define OPDESC_MASK_D12  OPDESC_MAKE(0xF, 0x1FF, 0x1FF, 0)
#define OPDESC_MASK_D1   OPDESC_MAKE(0xF, 0x1FF, 0,     0)
#define OPDESC_MASK_1    OPDESC_MAKE(0,   0x1FF, 0,     0)
#define OPDESC_MASK_12   OPDESC_MAKE(0,   0x1FF, 0x1FF, 0)

enum
{
	COND_EQ = 0,
	COND_NE,
	COND_LT,
	COND_LE,
	COND_GT,
	COND_GE,
};

//-----------------------------------------------------------------------------
// Global data
//-----------------------------------------------------------------------------

// Output buffer
#define MAX_VSH_SIZE 512
typedef std::vector<u32> outputBufType;
typedef outputBufType::iterator outputBufIter;
extern outputBufType g_outputBuf;

enum
{
	SE_PROC,
	SE_FOR,
	SE_IF,
	SE_ARRAY,
};

struct StackEntry
{
	int type;
	size_t pos;
	union
	{
		const char* strExtra;
		size_t uExtra;
	};
};

// Stack used to keep track of stuff.
#define MAX_STACK 32
extern StackEntry g_stack[MAX_STACK];
extern int g_stackPos;

// Operand descriptor stuff.
#define MAX_OPDESC 128
extern int g_opdescTable[MAX_OPDESC];
extern int g_opdeskMasks[MAX_OPDESC]; // used to keep track of used bits
extern int g_opdescCount;

enum
{
	UTYPE_BOOL = 0,
	UTYPE_IVEC,
	UTYPE_FVEC,
};

struct Uniform
{
	std::string name;
	int pos, size;
	int type;

	inline bool operator <(const Uniform& rhs) const
	{
		return pos < rhs.pos;
	}

	void init(const char* name, int pos, int size, int type)
	{
		this->name = name;
		this->pos = pos;
		this->size = size;
		this->type = type;
	}
};

// List of uniforms
#define MAX_UNIFORM 0x60
extern Uniform g_uniformTable[MAX_UNIFORM];
extern int g_uniformCount;

struct DVLEData; // Forward declaration

typedef std::pair<size_t, size_t> procedure; // position, size
typedef std::pair<size_t, std::string> relocation; // position, name

typedef std::map<std::string, procedure> procTableType;
typedef std::map<std::string, size_t> labelTableType;
typedef std::map<std::string, int> aliasTableType;
typedef std::vector<relocation> relocTableType;
typedef std::list<DVLEData> dvleTableType;

typedef procTableType::iterator procTableIter;
typedef labelTableType::iterator labelTableIter;
typedef aliasTableType::iterator aliasTableIter;
typedef relocTableType::iterator relocTableIter;
typedef dvleTableType::iterator dvleTableIter;

extern procTableType g_procTable;
extern dvleTableType g_dvleTable;
extern relocTableType g_procRelocTable;
extern int g_totalDvleCount;

// The following are cleared before each file is processed
extern labelTableType g_labels;
extern relocTableType g_labelRelocTable;
extern aliasTableType g_aliases;

extern bool g_autoNop;

int AssembleString(char* str, const char* initialFilename);
int RelocateProduct(void);

//-----------------------------------------------------------------------------
// Local data
//-----------------------------------------------------------------------------

enum
{
	OUTTYPE_POS      = 0,
	OUTTYPE_NQUAT    = 1,
	OUTTYPE_CLR      = 2,
	OUTTYPE_TCOORD0  = 3,
	OUTTYPE_TCOORD0W = 4,
	OUTTYPE_TCOORD1  = 5,
	OUTTYPE_TCOORD2  = 6,
	OUTTYPE_VIEW     = 8,
	OUTTYPE_DUMMY    = 9,
};

enum
{
	GSHTYPE_POINT    = 0,
	GSHTYPE_VARIABLE = 1,
	GSHTYPE_FIXED    = 2,
};

struct Constant
{
	int regId;
	int type;
	union
	{
		float fparam[4];
		u8 iparam[4];
		bool bparam;
	};
};

struct DVLEData
{
	// General config
	std::string filename;
	std::string entrypoint;
	size_t entryStart, entryEnd;
	bool nodvle, isGeoShader, isCompatGeoShader, isMerge;
	u16 inputMask, outputMask;
	u8 geoShaderType;
	u8 geoShaderFixedStart;
	u8 geoShaderVariableNum;
	u8 geoShaderFixedNum;

	// Uniforms
	Uniform uniformTable[MAX_UNIFORM];
	int uniformCount;
	size_t symbolSize;

	// Constants
	#define MAX_CONSTANT 0x60
	Constant constantTable[MAX_CONSTANT];
	int constantCount;

	// Outputs
	#define MAX_OUTPUT 16
	u64 outputTable[MAX_OUTPUT];
	u32 outputUsedReg;
	int outputCount;

	bool usesGshSpace() const { return isGeoShader && !isCompatGeoShader; }
	int findFreeOutput() const
	{
		for (int i = 0; i < maxOutputReg(); i ++)
			if (!(outputMask & BIT(i)))
				return i;
		return -1;
	}

	int findFreeInput() const
	{
		for (int i = 0; i < 16; i ++)
			if (!(inputMask & BIT(i)))
				return i;
		return -1;
	}

	int maxOutputReg() const
	{
		return isGeoShader ? 0x07 : 0x10;
	}

	DVLEData(const char* filename) :
		filename(filename), entrypoint("main"),
		nodvle(false), isGeoShader(false), isCompatGeoShader(false), isMerge(false),
		inputMask(0), outputMask(0), geoShaderType(0), geoShaderFixedStart(0), geoShaderVariableNum(0), geoShaderFixedNum(0),
		uniformCount(0), symbolSize(0), constantCount(0), outputUsedReg(0), outputCount(0) { }
};


source/FileClass.h
#pragma once
#include <stdio.h>
#include "types.h"

class FileClass
{
	FILE* f;
	bool LittleEndian, own;
	int filePos;

	size_t _RawRead(void* buffer, size_t size)
	{
		size_t x = fread(buffer, 1, size, f);
		filePos += x;
		return x;
	}

	size_t _RawWrite(const void* buffer, size_t size)
	{
		size_t x = fwrite(buffer, 1, size, f);
		filePos += x;
		return x;
	}

public:
	FileClass(const char* file, const char* mode) : LittleEndian(true), own(true), filePos(0)
	{
		f = fopen(file, mode);
	}
	FileClass(FILE* inf) : f(inf), LittleEndian(true), own(false), filePos(0) { }
	~FileClass()
	{
		if (f && own) fclose(f);
	}

	void SetLittleEndian() { LittleEndian = true; }
	void SetBigEndian() { LittleEndian = false; }

	FILE* get_ptr() { return f; }
	bool openerror() { return f == NULL; }

	dword_t ReadDword()
	{
		dword_t value;
		_RawRead(&value, sizeof(dword_t));
		return LittleEndian ? le_dword(value) : be_dword(value);
	}

	void WriteDword(dword_t value)
	{
		value = LittleEndian ? le_dword(value) : be_dword(value);
		_RawWrite(&value, sizeof(dword_t));
	}

	word_t ReadWord()
	{
		word_t value;
		_RawRead(&value, sizeof(word_t));
		return LittleEndian ? le_word(value) : be_word(value);
	}

	void WriteWord(word_t value)
	{
		value = LittleEndian ? le_word(value) : be_word(value);
		_RawWrite(&value, sizeof(word_t));
	}

	hword_t ReadHword()
	{
		hword_t value;
		_RawRead(&value, sizeof(hword_t));
		return LittleEndian ? le_hword(value) : be_hword(value);
	}

	void WriteHword(hword_t value)
	{
		value = LittleEndian ? le_hword(value) : be_hword(value);
		_RawWrite(&value, sizeof(hword_t));
	}

	byte_t ReadByte()
	{
		byte_t value;
		_RawRead(&value, sizeof(byte_t));
		return value;
	}

	void WriteByte(byte_t value)
	{
		_RawWrite(&value, sizeof(byte_t));
	}

	float ReadFloat()
	{
		union { word_t w; float f; } t;
		t.w = ReadWord();
		return t.f;
	}

	void WriteFloat(float value)
	{
		union { word_t w; float f; } t;
		t.f = value;
		WriteWord(t.w);
	}
	
	bool ReadRaw(void* buffer, size_t size) { return _RawRead(buffer, size) == size; }
	bool WriteRaw(const void* buffer, size_t size) { return _RawWrite(buffer, size) == size; }

	void Seek(int pos, int mode) { fseek(f, pos, mode); }
	int Tell() { return filePos /*ftell(f)*/; }
	void Flush() { fflush(f); }
};

static inline char* StringFromFile(const char* filename)
{
	FILE* f = fopen(filename, "rb");
	if (!f) return NULL;
	fseek(f, 0, SEEK_END);
	int size = ftell(f);
	rewind(f);
	char* buf = (char*)malloc(size+1);
	if (!buf)
	{
		fclose(f);
		return NULL;
	}
	fread(buf, 1, size, f);
	buf[size] = 0;
	fclose(f);
	return buf;
}


source/maestro_opcodes.h
#pragma once
enum
{
	MAESTRO_ADD = 0x00,
	MAESTRO_DP3,
	MAESTRO_DP4,
	MAESTRO_DPH,
	MAESTRO_DST,
	MAESTRO_EX2,
	MAESTRO_LG2,
	MAESTRO_LITP,
	MAESTRO_MUL,
	MAESTRO_SGE,
	MAESTRO_SLT,
	MAESTRO_FLR,
	MAESTRO_MAX,
	MAESTRO_MIN,
	MAESTRO_RCP,
	MAESTRO_RSQ,

	MAESTRO_unk10,
	MAESTRO_unk11,
	MAESTRO_MOVA,
	MAESTRO_MOV,
	MAESTRO_unk14,
	MAESTRO_unk15,
	MAESTRO_unk16,
	MAESTRO_unk17,
	MAESTRO_DPHI,
	MAESTRO_DSTI,
	MAESTRO_SGEI,
	MAESTRO_SLTI,
	MAESTRO_unk1C,
	MAESTRO_unk1D,
	MAESTRO_unk1E,
	MAESTRO_unk1F,

	MAESTRO_BREAK,
	MAESTRO_NOP,
	MAESTRO_END,
	MAESTRO_BREAKC,
	MAESTRO_CALL,
	MAESTRO_CALLC,
	MAESTRO_CALLU,
	MAESTRO_IFU,
	MAESTRO_IFC,
	MAESTRO_FOR,
	MAESTRO_EMIT, // Geometry shader related
	MAESTRO_SETEMIT, // Geometry shader related
	MAESTRO_JMPC,
	MAESTRO_JMPU,
	MAESTRO_CMP, // only the upper 5 bits are used for the opcode

	// Only the upper 3 bits are used for the following opcodes
	MAESTRO_MADI = 0x30,
	MAESTRO_MAD = 0x38,
};


source/picasso_frontend.cpp
#include "picasso.h"

// f24 has:
//  - 1 sign bit
//  - 7 exponent bits
//  - 16 mantissa bits
uint32_t f32tof24(float f)
{
	uint32_t i;
	memcpy(&i, &f, sizeof(f));

	uint32_t mantissa = (i << 9) >>  9;
	int32_t  exponent = (i << 1) >> 24;
	uint32_t sign     = (i << 0) >> 31;

	// Truncate mantissa
	mantissa >>= 7;

	// Re-bias exponent
	exponent = exponent - 127 + 63;
	if (exponent < 0)
	{
		// Underflow: flush to zero
		return sign << 23;
	}
	else if (exponent > 0x7F)
	{
		// Overflow: saturate to infinity
		return (sign << 23) | (0x7F << 16);
	}

	return (sign << 23) | (exponent << 16) | mantissa;
}

#ifdef WIN32
static inline void FixMinGWPath(char* buf)
{
	if (buf && *buf == '/')
	{
		buf[0] = buf[1];
		buf[1] = ':';
	}
}
#endif

int usage(const char* prog)
{
	fprintf(stderr,
		"Usage: %s [options] files...\n"
		"Options:\n"
		"  -o, --out=<file>        Specifies the name of the SHBIN file to generate\n"
		"  -h, --header=<file>     Specifies the name of the header file to generate\n"
		"  -n, --no-nop            Disables the automatic insertion of padding NOPs\n"
		"  -v, --version           Displays version information\n"
		, prog);
	return EXIT_FAILURE;
}

int main(int argc, char* argv[])
{
	char *shbinFile = NULL, *hFile = NULL;

	static struct option long_options[] =
	{
		{ "out",    required_argument, NULL, 'o' },
		{ "header", required_argument, NULL, 'h' },
		{ "help",   no_argument,       NULL, '?' },
		{ "no-nop", no_argument,       NULL, 'n' },
		{ "version",no_argument,       NULL, 'v' },
		{ NULL, 0, NULL, 0 }
	};

	int opt, optidx = 0;
	while ((opt = getopt_long(argc, argv, "o:h:?nv", long_options, &optidx)) != -1)
	{
		switch (opt)
		{
			case 'o': shbinFile = optarg; break;
			case 'h': hFile     = optarg; break;
			case '?': usage(argv[0]); return EXIT_SUCCESS;
			case 'n': g_autoNop = false; break;
			case 'v': printf("%s - Built on %s %s\n", PACKAGE_STRING, __DATE__, __TIME__); return EXIT_SUCCESS;
			default:  return usage(argv[0]);
		}
	}

#ifdef WIN32
	FixMinGWPath(shbinFile);
	FixMinGWPath(hFile);
#endif

	if (optind == argc)
	{
		fprintf(stderr, "%s: no input files are specified\n", argv[0]);
		return usage(argv[0]);
	}

	if (!shbinFile)
	{
		fprintf(stderr, "%s: no output file is specified\n", argv[0]);
		return usage(argv[0]);
	}

	int rc = 0;
	for (int i = optind; i < argc; i ++)
	{
		char* vshFile = argv[i];

#ifdef WIN32
		FixMinGWPath(vshFile);
#endif

		char* sourceCode = StringFromFile(vshFile);
		if (!sourceCode)
		{
			fprintf(stderr, "error: cannot open input file: %s\n", vshFile);
			return EXIT_FAILURE;
		}

		rc = AssembleString(sourceCode, vshFile);
		free(sourceCode);
		if (rc != 0)
			return EXIT_FAILURE;
	}

	rc = RelocateProduct();
	if (rc != 0)
		return EXIT_FAILURE;

	FileClass f(shbinFile, "wb");

	if (f.openerror())
	{
		fprintf(stderr, "Can't open output file!");
		return EXIT_FAILURE;
	}

	u32 progSize = g_outputBuf.size();
	u32 dvlpSize = 10*4 + progSize*4 + g_opdescCount*8;

	// Write DVLB header
	f.WriteWord(0x424C5644); // DVLB
	f.WriteWord(g_totalDvleCount); // Number of DVLEs

	// Calculate and write DVLE offsets
	u32 curOff = 2*4 + g_totalDvleCount*4 + dvlpSize;
	for (dvleTableIter dvle = g_dvleTable.begin(); dvle != g_dvleTable.end(); ++dvle)
	{
		if (dvle->nodvle) continue;
		f.WriteWord(curOff);
		curOff += 16*4; // Header
		curOff += dvle->constantCount*20;
		curOff += dvle->outputCount*8;
		curOff += dvle->uniformCount*8;
		curOff += dvle->symbolSize;
		curOff  = (curOff + 3) &~ 3; // Word alignment
	}

	// Write DVLP header
	f.WriteWord(0x504C5644); // DVLP
	f.WriteWord(0); // version
	f.WriteWord(10*4); // offset to shader binary blob
	f.WriteWord(progSize); // size of shader binary blob
	f.WriteWord(10*4 + progSize*4); // offset to opdesc table
	f.WriteWord(g_opdescCount); // number of opdescs
	f.WriteWord(dvlpSize); // offset to symtable (TODO)
	f.WriteWord(0); // ????
	f.WriteWord(0); // ????
	f.WriteWord(0); // ????

	// Write program
	for (outputBufIter it = g_outputBuf.begin(); it != g_outputBuf.end(); ++it)
		f.WriteWord(*it);

	// Write opdescs
	for (int i = 0; i < g_opdescCount; i ++)
		f.WriteDword(g_opdescTable[i]);

	// Write DVLEs
	for (dvleTableIter dvle = g_dvleTable.begin(); dvle != g_dvleTable.end(); ++dvle)
	{
		if (dvle->nodvle) continue;
		curOff = 16*4;

		f.WriteWord(0x454C5644); // DVLE
		f.WriteHword(0x1002); // maybe version?
		f.WriteByte(dvle->isGeoShader ? 1 : 0); // Shader type
		f.WriteByte(dvle->isMerge ? 1 : 0);
		f.WriteWord(dvle->entryStart); // offset to main
		f.WriteWord(dvle->entryEnd); // offset to end of main
		f.WriteHword(dvle->inputMask);
		f.WriteHword(dvle->outputMask);
		f.WriteByte(dvle->geoShaderType);
		f.WriteByte(dvle->geoShaderFixedStart);
		f.WriteByte(dvle->geoShaderVariableNum);
		f.WriteByte(dvle->geoShaderFixedNum);
		f.WriteWord(curOff); // offset to constant table
		f.WriteWord(dvle->constantCount); // size of constant table
		curOff += dvle->constantCount*5*4;
		f.WriteWord(curOff); // offset to label table (TODO)
		f.WriteWord(0); // size of label table (TODO)
		f.WriteWord(curOff); // offset to output table
		f.WriteWord(dvle->outputCount); // size of output table
		curOff += dvle->outputCount*8;
		f.WriteWord(curOff); // offset to uniform table
		f.WriteWord(dvle->uniformCount); // size of uniform table
		curOff += dvle->uniformCount*8;
		f.WriteWord(curOff); // offset to symbol table
		f.WriteWord(dvle->symbolSize); // size of symbol table

		// Sort uniforms by position
		std::sort(dvle->uniformTable, dvle->uniformTable + dvle->uniformCount);

		// Write constants
		for (int i = 0; i < dvle->constantCount; i ++)
		{
			Constant& ct = dvle->constantTable[i];
			f.WriteHword(ct.type);
			if (ct.type == UTYPE_FVEC)
			{
				f.WriteHword(ct.regId-0x20);
				for (int j = 0; j < 4; j ++)
					f.WriteWord(f32tof24(ct.fparam[j]));
			} else if (ct.type == UTYPE_IVEC)
			{
				f.WriteHword(ct.regId-0x80);
				for (int j = 0; j < 4; j ++)
					f.WriteByte(ct.iparam[j]);
			} else if (ct.type == UTYPE_BOOL)
			{
				f.WriteHword(ct.regId-0x88);
				f.WriteWord(ct.bparam ? 1 : 0);
			}
			if (ct.type != UTYPE_FVEC)
				for (int j = 0; j < 3; j ++)
					f.WriteWord(0); // Padding
		}

		// Write outputs
		for (int i = 0; i < dvle->outputCount; i ++)
			f.WriteDword(dvle->outputTable[i]);

		// Write uniforms
		size_t sp = 0;
		for (int i = 0; i < dvle->uniformCount; i ++)
		{
			Uniform& u = dvle->uniformTable[i];
			size_t l = u.name.length()+1;
			f.WriteWord(sp); sp += l;
			int pos = u.pos;
			if (pos >= 0x20)
				pos -= 0x10;
			f.WriteHword(pos);
			f.WriteHword(pos+u.size-1);
		}

		// Write symbols
		for (int i = 0; i < dvle->uniformCount; i ++)
		{
			std::string u(dvle->uniformTable[i].name);
			std::replace(u.begin(), u.end(), '$', '.');
			size_t l = u.length()+1;
			f.WriteRaw(u.c_str(), l);
		}

		// Word alignment
		int pos = f.Tell();
		int pad = ((pos+3)&~3)-pos;
		for (int i = 0; i < pad; i ++)
			f.WriteByte(0);
	}

	if (hFile)
	{
		FILE* f2 = fopen(hFile, "w");
		if (!f2)
		{
			fprintf(stderr, "Can't open header file!\n");
			return 1;
		}

		fprintf(f2, "// Generated by picasso\n");
		fprintf(f2, "#pragma once\n");
		const char* prefix = g_dvleTable.front().isGeoShader ? "GSH" : "VSH"; // WARNING: HORRIBLE HACK - PLEASE FIX!!!!!!!
		for (int i = 0; i < g_uniformCount; i ++)
		{
			Uniform& u = g_uniformTable[i];
			const char* name = u.name.c_str();
			if (*name == '_') continue; // Hidden uniform
			if (u.type == UTYPE_FVEC)
				fprintf(f2, "#define %s_FVEC_%s 0x%02X\n", prefix, name, u.pos-0x20);
			else if (u.type == UTYPE_IVEC)
				fprintf(f2, "#define %s_IVEC_%s 0x%02X\n", prefix, name, u.pos-0x80);
			else if (u.type == UTYPE_BOOL)
			{
				if (u.size == 1)
					fprintf(f2, "#define %s_FLAG_%s BIT(%d)\n", prefix, name, u.pos-0x88);
				else
					fprintf(f2, "#define %s_FLAG_%s(_n) BIT(%d+(_n))\n", prefix, name, u.pos-0x88);
			}
			fprintf(f2, "#define %s_ULEN_%s %d\n", prefix, name, u.size);
		}

		fclose(f2);
	}

	return EXIT_SUCCESS;
}
